### ISSUE TITLE
esc directly after creation of an auto-indented line does not remove whitespace

### ISSUE THREAD

**Issue by davea1** [NeoVintageous/NeoVintageous#440]
While typing text in insert mode with the `auto_indent` setting enabled, `esc` immediately following the creation of a new line (via `Enter`/`Return`) will leave behind whitespace.

With Vim, all such whitespace is removed as  soon as `esc` is hit, leaving only a blank line.

This issue is related to: #380

**Comment by doei** [comment]
I cannot tell the difference between this issue and #380 but this one's opened and the other is closed so I'll comment here.

I came here with the intention to report the exact behaviour described by this issue because I'd like to have it back (I would very much like the whitespace to stay where it is) but I suppose it turns out to be something that does not comply with vim's specs, and thus has been corrected since then.

Would you consider allowing for the old behaviour under some setting ?

Here's my use-case if that's interesting: from anywhere on a line I like to be able to hit `o<Esc>hp` and have my stuff pasted on a new line under the previous one, and at the correct indentation level.

Thanks for considering :)

**Comment by gerardroche** [comment]
Yeah, removing the leading whitespace on Esc seems consistent with Vim: https://vi.stackexchange.com/questions/1873/why-does-the-cursor-move-to-the-start-of-the-line-after-esc

In the SO post it says you can press `<BS><Esc>` instead of just `<Esc>` to preserve the whitespace, but I couldn't get Vim to do that.

In Vim, the indentation is controlled by `'autoindent'`, which is the same as the Sublime `auto_indent` setting. 

Pasting using indentation in Vim seem to be achieved by `[p`, `[P`,  `]p` and `]P` #339. For example, if you want to paste above the current line, instead of pressing `O<Esc>p` you just press `[P` or `[p`. And if you want to paste below `]P` or `]p`.

It it would be better to implement `]P` rather than adding a setting, but adding a setting is not a big a deal, e.g. `'vintageous_clear_auto_indent_on_esc'` (default true).

**Comment by doei** [comment]
brackets aren't super easy to use on my keyboard I have to say, but it'd be better to have a solution for it rather than nothing, otherwise that setting would be great of course. Are you saying you're considering it ?

**Comment by gerardroche** [comment]
Yes, I'll add the setting. I might add it in the next patch rather than the next minor version.

**Comment by gerardroche** [comment]
@doei I'll be cutting 1.15.1, which has the new setting, in a few minutes so it should be available to upgrade in Package Control shortly.

**Comment by davea1** [comment]
> In the SO post it says you can press `<BS><Esc>` instead of just `<Esc>` to preserve the whitespace, but I couldn't get Vim to do that.

It was noted in the SO thread that in Vim you need to type a character before `<BS><Esc>` in order to preserve the whitespace. I have verified that this appears to work for me in my Vim instances.

Regarding the new setting, it is always great to have choices! :)

@gerardroche Any estimate on when this specific issue (#440) might be addressed? Thanks.

**Comment by gerardroche** [comment]
>     In the SO post it says you can press <BS><Esc> instead of just <Esc> to preserve the whitespace, but I couldn't get Vim to do that.
> 
> It was noted in the SO thread that in Vim you need to type a character before <BS><Esc> in order to preserve the whitespace. I have verified that this appears to work for me in my Vim instances.

The way I read it was that you only had to type `<BS><Esc>` rather than `{char}<BS><Esc>`, which seems silly now that I think about it. In this case NeoVintageous works the same way.

> @gerardroche Any estimate on when this specific issue (#440) might be addressed? Thanks.

It will be a long while before this is fully resolved because it may require a lot of internals redesign or at least a better understanding of how the sublime apis work, and their limitations.

**Comment by davea1** [comment]
@gerardroche Thanks for the update! I hope this ends up being easier than anticipated. :)

I'm not sure if it will help, but you might examine how this issue was approached in the Vim extension for VSCode (https://github.com/VSCodeVim/Vim).

Note: Oddly, VSCodeVim currently deviates from proper Vim emulation in this respect when VSCode has been set to 'Text Mode', but appears to work correctly when the Language Mode has been set for editing code.
